src/mds/CInode.cc

```diff
@@ -5398,7 +5398,7 @@ void CInode::queue_export_pin(mds_rank_t target)
 
 void CInode::maybe_export_pin(bool update)
 {
-  if (!g_conf()->mds_bal_export_pin)
+  if (!mdcache->mds->balancer->get_bal_export_pin())
```

Comment:
@batrick I decided to use the cached value from MDBalancer here in CInode code but this change causes a build failure:
```
/home/sidharthanup/ceph/src/mds/CInode.cc: In member function ‘void CInode::maybe_export_pin(bool)’:
/home/sidharthanup/ceph/src/mds/CInode.cc:5401:30: error: invalid use of incomplete type ‘class MDBalancer’
   if (!mdcache->mds->balancer->get_bal_export_pin())
                              ^~
In file included from /home/sidharthanup/ceph/src/mds/CInode.cc:22:
/home/sidharthanup/ceph/src/mds/CDir.h:656:16: note: forward declaration of ‘class MDBalancer’
   friend class MDBalancer;
                ^~~~~~~~~~
make[2]: *** [src/mds/CMakeFiles/mds.dir/build.make:310: src/mds/CMakeFiles/mds.dir/CInode.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:5686: src/mds/CMakeFiles/mds.dir/all] Error 2
```
I didn't exactly understand the error stems from the declaration of the MDBalancer class in CDir.h?

Reply:
No, the MDBalancer class is incomplete because `MDBalancer.h` is not included in `CInode.cc`. You can either include it (probably not desirable for code hygiene reasons) or move the config caching to `MDSRank.h`